File: examples/python/mongodb/mongodb_fuse.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import os
import sys
import time

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.mongodb import MongoDBConfig, MongoDBResource

load_dotenv(".env.development")

DB = "mirage_test"
COLL = "heterogeneous"
VIEW = "high_rated_films"

config = MongoDBConfig(uri=os.environ["MONGODB_URI"], databases=[DB])
resource = MongoDBResource(config=config)

with Workspace(
    {"/mongodb/": resource},
        mode=MountMode.READ,
        fuse=True,
) as ws:
    time.sleep(1)
    mp = ws.fuse_mountpoint

    print(f"=== FUSE MODE: mounted at {mp} ===\n")

    print("--- listdir() root (databases) ---")
    for db in os.listdir(f"{mp}/mongodb"):
        print(f"  {db}")

    print(f"\n--- listdir() /mongodb/{DB} (entities) ---")
    for entry in os.listdir(f"{mp}/mongodb/{DB}"):
        print(f"  {entry}")

    print(f"\n--- listdir() /mongodb/{DB}/collections ---")
    for col in os.listdir(f"{mp}/mongodb/{DB}/collections"):
        print(f"  {col}")

    print(f"\n--- listdir() /mongodb/{DB}/views ---")
    for v in os.listdir(f"{mp}/mongodb/{DB}/views"):
        print(f"  {v}")

    print(f"\n--- listdir() /mongodb/{DB}/collections/{COLL} (entity) ---")
    for entry in os.listdir(f"{mp}/mongodb/{DB}/collections/{COLL}"):
        print(f"  {entry}")

    print("\n--- open() database.json ---")
    with open(f"{mp}/mongodb/{DB}/database.json") as f:
        db_meta = json.loads(f.read())
    print(f"  collections: {len(db_meta.get('collections', []))}")
    print(f"  views: {len(db_meta.get('views', []))}")

    print(f"\n--- open() schema.json for {COLL} ---")
    with open(f"{mp}/mongodb/{DB}/collections/{COLL}/schema.json") as f:
        schema = json.loads(f.read())
    print(f"  kind: {schema.get('kind')}")
    print(f"  fields: {len(schema.get('fields', []))}")
    print(f"  indexes: {len(schema.get('indexes', []))}")

    print(f"\n--- open() + read documents.jsonl for {COLL} ---")
    path = f"{mp}/mongodb/{DB}/collections/{COLL}/documents.jsonl"
    with open(path) as f:
        text = f.read().strip()
    lines = [ln for ln in text.splitlines() if ln.strip()]
    print(f"  documents: {len(lines)}")
    for line in lines[:3]:
        doc = json.loads(line)
        print(f"  {doc.get('title', '?')}")

    print(f"\n--- open() + read view documents for {VIEW} ---")
    view_path = f"{mp}/mongodb/{DB}/views/{VIEW}/documents.jsonl"
    with open(view_path) as f:
        view_text = f.read().strip()
    view_lines = [ln for ln in view_text.splitlines() if ln.strip()]
    print(f"  documents: {len(view_lines)}")
    for line in view_lines[:3]:
        doc = json.loads(line)
        print(f"  {doc.get('title', '?')} (rating={doc.get('rating', '?')})")

    print(f"\n>>> FUSE mounted at: {mp}")
    print(">>> Open another terminal and try:")
    print(f">>>   ls {mp}/mongodb/{DB}/collections/")
    print(
        f">>>   head -n 3 {mp}/mongodb/{DB}/collections/{COLL}/documents.jsonl"
    )
    if sys.stdin.isatty():
        print(">>> Press Enter to unmount and exit...")
        input()
    else:
        print(">>> (non-interactive: unmounting now)")

    records = ws.ops.records
    total = sum(r.bytes for r in records)
    print(f"\nStats: {len(records)} ops, "
          f"{total} bytes transferred")

File: examples/typescript/mongodb/mongodb_fuse.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { setServers } from "node:dns";
import { readdir, readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";
import dotenv from "dotenv";
import {
  MongoDBResource,
  Mount,
  MountMode,
  Workspace,
} from "@struktoai/mirage-node";

const __HERE = fileURLToPath(new URL(".", import.meta.url));
setServers(["8.8.8.8", "1.1.1.1"]);
dotenv.config({ path: resolve(__HERE, "../../../.env.development") });

const DB = "mirage_test";
const COLL = "heterogeneous";
const VIEW = "high_rated_films";

async function main(): Promise<void> {
  const uri = process.env.MONGODB_URI;
  if (uri === undefined) {
    console.error("MONGODB_URI missing in .env.development");
    process.exit(1);
  }
  const resource = new MongoDBResource({ uri, databases: [DB] });
  const ws = new Workspace({
    "/mongodb/": new Mount(resource, { mode: MountMode.READ, fuse: true }),
  });
  await ws.fuseReady();
  const mp = ws.fuseMountpoint as string;

  try {
    console.log(`=== FUSE MODE: mounted at ${mp} ===\n`);

    console.log("--- listdir() root (databases) ---");
    for (const db of (await readdir(mp)).sort()) console.log(`  ${db}`);

    console.log(`\n--- listdir() /mongodb/${DB} (entities) ---`);
    for (const e of (await readdir(`${mp}/${DB}`)).sort())
      console.log(`  ${e}`);

    console.log(`\n--- listdir() /mongodb/${DB}/collections ---`);
    for (const c of (await readdir(`${mp}/${DB}/collections`)).sort())
      console.log(`  ${c}`);

    console.log(`\n--- listdir() /mongodb/${DB}/views ---`);
    for (const v of (await readdir(`${mp}/${DB}/views`)).sort())
      console.log(`  ${v}`);

    console.log(
      `\n--- listdir() /mongodb/${DB}/collections/${COLL} (entity) ---`,
    );
    for (const e of (await readdir(`${mp}/${DB}/collections/${COLL}`)).sort())
      console.log(`  ${e}`);

    console.log("\n--- open() database.json ---");
    const dbMeta = JSON.parse(
      await readFile(`${mp}/${DB}/database.json`, "utf8"),
    ) as {
      collections: unknown[];
      views: unknown[];
    };
    console.log(`  collections: ${String(dbMeta.collections.length)}`);
    console.log(`  views: ${String(dbMeta.views.length)}`);

    console.log(`\n--- open() schema.json for ${COLL} ---`);
    const schema = JSON.parse(
      await readFile(`${mp}/${DB}/collections/${COLL}/schema.json`, "utf8"),
    ) as { kind: string; fields: unknown[]; indexes: unknown[] };
    console.log(`  kind: ${schema.kind}`);
    console.log(`  fields: ${String(schema.fields.length)}`);
    console.log(`  indexes: ${String(schema.indexes.length)}`);

    console.log(`\n--- open() + read documents.jsonl for ${COLL} ---`);
    const text = (
      await readFile(`${mp}/${DB}/collections/${COLL}/documents.jsonl`, "utf8")
    ).trim();
    const lines = text.split("\n").filter((ln) => ln.trim() !== "");
    console.log(`  documents: ${String(lines.length)}`);
    for (const ln of lines.slice(0, 3)) {
      const doc = JSON.parse(ln) as { title?: string };
      console.log(`  ${doc.title ?? "?"}`);
    }

    console.log(`\n--- open() + read view documents for ${VIEW} ---`);
    const viewText = (
      await readFile(`${mp}/${DB}/views/${VIEW}/documents.jsonl`, "utf8")
    ).trim();
    const viewLines = viewText.split("\n").filter((ln) => ln.trim() !== "");
    console.log(`  documents: ${String(viewLines.length)}`);
    for (const ln of viewLines.slice(0, 3)) {
      const doc = JSON.parse(ln) as { title?: string; rating?: number };
      console.log(
        `  ${doc.title ?? "?"} (rating=${String(doc.rating ?? "?")})`,
      );
    }

    console.log(`\n>>> FUSE mounted at: ${mp}`);
    console.log(">>> Open another terminal and try:");
    console.log(`>>>   ls ${mp}/${DB}/collections/`);
    console.log(
      `>>>   head -n 3 ${mp}/${DB}/collections/${COLL}/documents.jsonl`,
    );
    if (process.stdin.isTTY) {
      console.log(">>> Press Enter to unmount and exit...");
      await new Promise<void>((resolve) => {
        process.stdin.once("data", () => resolve());
      });
    } else {
      console.log(">>> (non-interactive: unmounting now)");
    }
  } finally {
    await ws.close();
    await resource.close();
  }
}

await main();
